### PR TITLE
Speed up clear_task_instances by doing a single sql delete for TaskReschedule

### PR DESCRIPTION
**Pull request by yuqian90**
Clearing large number of tasks takes a long time. Most of the time is spent at this line in `clear_task_instances` (more than 95% time). This slowness sometimes causes the webserver to timeout because the `web_server_worker_timeout` is hit.

```python
        # Clear all reschedules related to the ti to clear
        session.query(TR).filter(
            TR.dag_id == ti.dag_id,
            TR.task_id == ti.task_id,
            TR.execution_date == ti.execution_date,
            TR.try_number == ti.try_number,
        ).delete()
```

This line is very slow because it's deleting `TaskReschedule` rows in a for loop one by one. 

This PR simply changes this code to delete `TaskReschedule` in a single sql query with a bunch of `OR` conditions. It's effectively doing the same, but now it's much faster. Simple profiling shows that it's at least seven times faster when deleting thousands of `TaskReschedule`.

### CODE REVIEW

**Review by potiuk — APPROVED** [review]
Looks good to me. @kaxil - can you take a look on that one as well? It looks like nice thing to cherry-pick to 2.0.1, very localized and seems to improve the case that is rather useful.

File: airflow/models/taskinstance.py
Comment (mik-laj) [inline airflow/models/taskinstance.py]: Is `tr_filter` always contain at least one element? I'm afraid that if `tr_filter` is empty, too many objects in the database will be deleted because ORM won't generate SQL query without any filters.

Comment (yuqian90) [reply airflow/models/taskinstance.py]: Thanks. That's a good point. We should defend against this. I updated according to your suggestion to not delete anything if `tr_filter` is empty.

**Comment by yuqian90** [comment]
Thanks @potiuk  and @mik-laj for the review. I updated the PR according to your comments.

Also, I noticed that doing `TR.__table__.delete().where(...)` seems to be slightly faster than what I was doing with `session.query(...).filter(...).delete()` by a small margin. So I changed the PR to use it instead.

File: airflow/models/taskinstance.py
Comment (turbaszek) [inline airflow/models/taskinstance.py]: Should we append `tis` and then use:
https://github.com/apache/airflow/blob/2a3960f359b1fcd5eed64fff472c4e463da74beb/airflow/models/taskinstance.py#L1987

to avoid code duplication?

Comment (yuqian90) [reply airflow/models/taskinstance.py]: Hi @turbaszek. I looked into `filter_for_tis()`. It specifically creates a filter for `TaskInstance`. In this case, we need to filter for `TaskReschedule` instead. Also, it does not include a condition for `try_number`, which is necessary to get `TaskReschedule` right. 

```python
        return or_(
            and_(
                TaskInstance.dag_id == ti.dag_id,
                TaskInstance.task_id == ti.task_id,
                TaskInstance.execution_date == ti.execution_date,
            )
            for ti in tis
        )
```

**Review by turbaszek — APPROVED** [review]

File: airflow/models/taskinstance.py
Comment (kaxil) [inline airflow/models/taskinstance.py]: this can go at the top of the file

File: airflow/models/taskinstance.py
Comment (kaxil) [inline airflow/models/taskinstance.py]:
```suggestion
```

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
Minor suggestions added, LGTM otherwise

File: airflow/models/taskinstance.py
Comment (kaxil) [reply airflow/models/taskinstance.py]: actually it is already at the top of file so not needed

```suggestion
```

**Comment by kaxil** [comment]
Pushed https://github.com/apache/airflow/pull/14048/commits/99a77f3e368f5ea798deb5b97b30f14dbc298c2e -- Let's wait for the CI, will merge after that, thanks @yuqian90

File: airflow/models/taskinstance.py
Comment (ashb) [inline airflow/models/taskinstance.py]: If these are all for the same dag (which it looks like they _should_ be given the API) then this query would be much quicker if `TR.dag_id = dag.dag_id` is pulled out of the loop so it doesn't appear so often.

(This was an optimization I made to `filter_for_tis`)

Comment (yuqian90) [reply airflow/models/taskinstance.py]: Thanks. That's a great idea. I looked into the optimization in `filter_for_tis`. It works by extracting the common `dag_id` and `execution_date`, and then filter the `task_id` with an `IN` statement. This greatly reduces the size of the query, but it only does so if both `dag_id` and `execution_date` are equal for all tis. 

For the case of slow `clear_task_instances`, the large number of tis is usually produced by the combination of a single dag_id over multiple `execution_date` for a large dag containing many task_ids. So it's slightly more complicated because there can be multiple `dag_id` and `execution_date` combinations (as well as `try_number`).
So I updated the logic to be more generic. It now uses a nested dictionary to produce the hierarchical filter. Some profiling showed great speed improvement (something like 40 to 50 times faster) compared to the first iteration. So the overral performance should now be 300 times faster than the original for loop deletion.

**Comment by yuqian90** [comment]
Followed @ashb 's [suggestion](https://github.com/apache/airflow/pull/14048#discussion_r571501423) and speed it up even further. Please take another look.

File: airflow/models/taskinstance.py
Comment (kaxil) [inline airflow/models/taskinstance.py]: We can move this in the loop above instead of looping across it again, right?

Comment (kaxil) [reply airflow/models/taskinstance.py]: Can you post the code that calculates times takes to see 40x speed

Comment (yuqian90) [reply airflow/models/taskinstance.py]: Oh. Thanks @kaxil you are exactly right! Updated accordingly.

**Comment by yuqian90** [comment]
Hi @kaxil this is the experiment I did. `test_create_large_dag_with_task_reschedule` creates a log of TaskInstance and `TaskReschedule` in the db. Then `test_clear_large_dag_with_task_reschedule` times how long it takes to call `clear_task_instances` on them.

```python
import pytest

@pytest.fixture(scope="module")
def big_dag():
    with DAG(
        'test_create_large_dag_with_task_reschedule',
        start_date=DEFAULT_DATE,
        end_date=DEFAULT_DATE + datetime.timedelta(days=30),
    ) as dag:
        for i in range(1000):
            PythonSensor(task_id=f'task_{i}', python_callable=lambda: False, mode="reschedule")

    yield dag

import copy

def test_create_large_dag_with_task_reschedule(big_dag):

    tis = []

    for i in range(10):
        execution_date = DEFAULT_DATE + datetime.timedelta(days=i)
        big_dag.create_dagrun(
            execution_date=execution_date,
            state=State.RUNNING,
            run_type=DagRunType.SCHEDULED,
        )
        for task in big_dag.tasks:
            ti = TI(task=copy.copy(task), execution_date=execution_date)
            tis.append(ti)

    import pendulum

    tss = []
    for ti in tis:
        for i in range(5):
            tss.append(TaskReschedule(
                task=ti.task,
                execution_date=ti.execution_date,
                try_number=1,
                start_date=pendulum.now(),
                end_date=pendulum.now(),
                reschedule_date=pendulum.now()))

    def chunks(lst, n):
        """Yield successive n-sized chunks from lst."""
        for i in range(0, len(lst), n):
            yield lst[i:i + n]


    with create_session() as session:
        for chunk in chunks([{"task_id": tr.task_id, "dag_id": tr.dag_id, "execution_date": tr.execution_date,
                              "try_number": tr.try_number, "start_date": tr.start_date, "end_date": tr.end_date,
                              "duration": tr.duration, "reschedule_date": tr.reschedule_date} for tr in tss], 10000):
            session.execute(TaskReschedule.__table__.insert(), chunk)
            session.commit()

def test_clear_large_dag_with_task_reschedule(big_dag):
    import time

    with create_session() as session:
        def count_task_reschedule():
            return (
                session.query(TaskReschedule)
                .filter(
                    TaskReschedule.dag_id == big_dag.dag_id,
                    TaskReschedule.try_number == 1,
                )
                .count()
            )

        assert count_task_reschedule() == 1000 * 10 * 5
        qry = session.query(TI).filter(TI.dag_id == big_dag.dag_id).all()
        start = time.time()
        clear_task_instances(qry, session, dag=big_dag)
        end = time.time()
        print(f"clear_task_instances took {end - start}")
        assert count_task_reschedule() == 0
        session.rollback()
```

File: airflow/models/taskinstance.py
Comment (ashb) [reply airflow/models/taskinstance.py]: Yeah -- the performance difference is crazy for "large" queries in cases like this.

**Pull request merged**
